rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc

```diff
@@ -0,0 +1,156 @@
+@license{
+Copyright (c) 2018-2023, NWO-I CWI and Swat.engineering
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+this list of conditions and the following disclaimer in the documentation
+and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+@bootstrapParser
+@synopsis{Defines both the command evaluator and the codeAction retriever for Rascal}
+module lang::rascal::lsp::Actions
+
+import lang::rascal::\syntax::Rascal;
+import util::LanguageServer;
+import analysis::diff::edits::TextEdits;
+import ParseTree;
+import String;
+import lang::rascal::vis::ImportGraph;
+import util::Reflective;
+import util::IDEServices;
+import List;
+import IO;
+
+@synopsis{Here we list Rascal-specific code commands}
+@description{
+The commands must be evaluated by ((evaluateRascalCommand))
+}
+data Command
+    = visualImportGraphCommand(PathConfig pcfg)
+    | sortImportsAndExtends(Header h)
+    ;
+
+@synopsis{Detects (on-demand) source actions to register with specific places near the current cursor}
+list[CodeAction] rascalCodeActions(Focus focus, PathConfig pcfg=pathConfig()) {
+    result = [];
+
+    if ([*_, start[Module] top] := focus) {
+        result += addLicenseAction(top, pcfg);
+    }
+
+    if ([*_, Toplevel t, *_] := focus) {
+        result += toplevelCodeActions(t);
+    }
+
+    if ([*_, Header h, *_] := focus) {
+        result += [action(command=visualImportGraphCommand(pcfg), title="Visualize project import graph")]
+               +  [action(command=sortImportsAndExtends(h), title="Sort imports and extends")]
```

Comment:
we should only offer this if the header is more than 1 import?

Reply:
It shows all modules in the entire project always, so it is not that specific to the current module.